### PR TITLE
add gpt and t5 model

### PR DESCRIPTION
**Pull request by dangkai4u**
- [x] 实现模型架构
- [x] 生成mask、position_ids操作
- [x] 配置系统
- [x] 损失函数

### CODE REVIEW

**Comment by dangkai4u** [comment]
mask_helper设计：将所有涉及attention mask的操作分为两类，一个是维度拓展，一个是添加casual mask。所以构造了2个算子，ExtendedMask和CasualMask，分别实现这两个功能，使用时分别声明并调用。

**Comment by dangkai4u** [comment]
gpt和t5实现了整体，细节上需要和megatron校对一下。感觉在模型输出上，可能需要统一一下格式，输出hidden states，还是输出logits，还是输出loss，这些需要统一。

**Comment by dangkai4u** [comment]
google的T5与普通Transformer的区别：
1. layernorm：相比于正常的layernorm，少了bias项和减mean的操作；
2. FFN层：激活函数为relu或者为gated gelu；
3. attention操作：使用相对位置编码，一种基于桶的相对位置编码
4. 取消了绝对位置编码
5. embedding共享

megatron的T5与普通Transformer的区别：
1. encoder和decoder的embedding共享
2. 其余没区别

现在把预训练模型分为BERT类、GPT类、T5类，分类依据是模型架构不同，BERT基于Transformer的Encoder，GPT基于Transformer的decoder，T5基于Transformer。huggingface下实现了很多种预训练模型，比如roberta、albert、bart、xlnet、unilm等，这些都可以基于上述三个架构，加上一些细节或任务上的改动完成。例如，roberta是bert经过更大的训练数据、精细化调整的模型；unilm在架构上和BERT相同，调整了attention mask，统一了MLM、LM、Seq2Seq模型；xlnet将预训练任务改为乱序语言模型；bart和T5相同，在具体实现和数据处理上存在细微差别。因此，当前阶段，我认为只需实现BERT、GPT、T5三类，且细节上可以与原始模型存在区别，重点是搭建出这三种架构的模型。

**Comment by L1aoXingyu** [comment]
建议 layers 里面的修改单独提一个 pr，review 之后就可以合并了
T5 和 gpt 分别开 PR，这样两个模型可以各自合并，不用全部放到一个 PR 里面

**Comment by dangkai4u** [comment]
layers修改不会影响目前的模型，只是写t5的时候发现，embedding层不一定都在第0层，对于transformer模型，有encoder和decoder分别输入input_ids，分别计算embedding。当不共享embedding模块时，放在不同的placement上可能会减少通信次数。

t5和gpt可以一起合并，如果验证完t5模型，bert和gpt都不用单独验证了，因为用的都是那些模块，t5相当于把所有路径都测试了一遍。如果暂时不验证推理的话，t5、gpt和bert所涉及的计算是一样的，也不用验证了。

**Comment by L1aoXingyu** [comment]
layers 的正确性也许可以保证，那模型的正确性也能保证吗？

**Comment by dangkai4u** [comment]
当然可以，只要搭建模型的人熟练，再多经过几次检查就行，模型的错可以通过检查发现，算子的错才需要框架间对齐。

**Comment by dangkai4u** [comment]
这个pr还有什么问题吗

File: libai/models/gpt_model.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+        return output
+
+
+class GPTEmbedding(nn.Module):
+    def __init__(
+        self, 
+        vocab_size, 
+        hidden_size, 
+        max_seq_length, 
+        init_method=init.xavier_normal_,
```
Comment (marigoold) [inline libai/models/gpt_model.py:135]: 这里`init`没有import

File: libai/models/gpt_model.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        )
+        self.layernorm_f = LayerNorm(hidden_size, eps=layernorm_epsilon, layer_idx=-1)
+
+
+    def forward(self, hidden_states, attention_mask, past_key_values=None, use_cache=False):
+        # hidden_states shape: (batch_size, seq_length, hidden_size)
+        # sbp: [S(0), B]
+        presents = []
+        if past_key_values is None:
+            past_key_values = tuple([None] * len(self.num_layers))
```
Comment (marigoold) [inline libai/models/gpt_model.py:210]:
```suggestion
            past_key_values = tuple([None] * len(self.layers))
```
这里有一个变量用错了

File: libai/data/datasets/bert_dataset.py [hunk 3]
```diff
@@ -121,7 +120,6 @@ def __getitem__(self, idx):
                 flow.tensor(int(is_next_random), dtype=flow.long), placement_idx=-1
             ),
             lm_labels=DistTensorData(labels, placement_idx=-1),
-            loss_mask=DistTensorData(loss_mask, placement_idx=-1),
```
Comment (dangkai4u) [inline libai/data/datasets/bert_dataset.py:124]: 去掉loss_mask，通过ignore_index进行掩码。这样做的好处是接口和pytorch的CrossEntropyLoss更一致，易用性更强。至于效率问题，之前是在数据处理中构造loss_mask，现在在损失计算时生成loss_mask，一个是cpu操作，一个是gpu操作，效率不会低太多，可能会有提升。

File: libai/layers/cross_entropy.py [hunk 4]
```diff
@@ -17,23 +17,43 @@
...
-    def forward(self, logits, target):
-        """Function for the distributed cross entropy
-        vocab_parallel_logits with shape (batch_size, seq_length, vocab_size)
-        and sbp sign [S(0), S(2)].
-        target with shape (batch_size, seq_length) and sbp sign [S(0), B].
-        """
-        assert logits.ndim == 3
-        assert target.ndim == 2
-        assert logits.shape[0:2] == target.shape
+    def __init__(self, ignore_index=None, reduction="mean"):
```
Comment (dangkai4u) [inline libai/layers/cross_entropy.py:24]: 重新设计交叉熵损失，api接口类似，仿照bert_loss的逻辑和函数重新封装，不会有新的逻辑和corner case错误，初步验证损失正常。之后有时间可以添加label_smoothing。

File: libai/layers/embedding.py [hunk 5]
```diff
@@ -39,8 +41,11 @@ def __init__(
         embedding_dim,
         padding_idx=None,
         init_method=init.xavier_normal_,
+        *,
+        layer_idx=0,
```
Comment (dangkai4u) [inline libai/layers/embedding.py:45]: 新增layer_idx参数，embedding模块不一定只存在于第一个设备，在transformer模型中，如果不使用embedding模块共享，即encoder和decoder分别有一个embedding模块，此时使用流水并行时，decoder的embedding模块放在decoder的第一层更合适，可以减少通信。

File: libai/layers/mask_helpers.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+import oneflow as flow
+import oneflow.nn as nn
+
+from libai.utils import distributed as dist
+
+
+class ExtendedMask(nn.Module):
+    """Makes the attention mask broadcastable at the head dims."""
+
+    def forward(self, attention_mask):
```
Comment (dangkai4u) [inline libai/layers/mask_helpers.py:25]: 将attention_mask拓展维度，transformer里attention_mask有以下几种情形：
1. 二维向量（batch_size, seq_length）：通过input_ids.ne(pad_idx)得到，用来表明哪些词是需要被关注，需将其拓展为（batch_size, 1, 1, seq_length）
2. 三维向量（batch_size, tgt_length, src_length）：用户构造，需将其拓展为（batch_size, 1, tgt_length, src_length）
3. 四维向量（batch_size, 1, tgt_length, src_length）

由于将其转化为int8类型，是因为fuse算子参数只支持int8，暂时不支持别的类型

File: libai/layers/mask_helpers.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+            # we extend it to [batch_size, 1, 1, src_len]
+            extended_attention_mask = attention_mask.unsqueeze(1).unsqueeze(2)
+        else:
+            raise ValueError(f"Wrong shape for attention_mask (shape {attention_mask.shape})")
+
+        extended_attention_mask = extended_attention_mask.to(flow.int8)
+        return extended_attention_mask
+
+
+class CasualMask(nn.Module):
```
Comment (dangkai4u) [inline libai/layers/mask_helpers.py:41]: 生成casual_mask，也就是下三角矩阵，attention_mask不为空时，将pad_mask和casual_mask合并。

File: libai/models/gpt_model.py [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+            "layernorm_epsilon": cfg.layernorm_epsilon,
+            "initializer_range": cfg.initializer_range,
+            "bias_gelu_fusion": cfg.bias_gelu_fusion,
+            "bias_dropout_fusion": cfg.bias_dropout_fusion,
+            "scale_mask_softmax_fusion": cfg.scale_mask_softmax_fusion,
+            "apply_query_key_layer_scaling": cfg.apply_query_key_layer_scaling,
+        }
+
+    def forward(self, input_ids, past_key_values=None, use_cache=False):
+        past_length = past_key_values[0][0].size(2) if past_key_values is not None else 0
```
Comment (dangkai4u) [inline libai/models/gpt_model.py:231]: past_key_values保存了num_layers个past_key_value，past_key_value在当前层计算attention时使用，减少重复计算。

File: libai/models/gpt_model.py [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+
+    def forward(
+        self,
+        input_ids,
+        labels=None,
+        past_key_values=None,
+        use_cache=False,
+    ):
+        outputs = self.gpt_model(input_ids, past_key_values=past_key_values, use_cache=use_cache)
+        if not isinstance(outputs, (tuple, list)):
```
Comment (dangkai4u) [inline libai/models/gpt_model.py:268]: 这里是因为oneflow的bug，做的重复操作。graph会自动unpack长度为1的元组。待该bug修复后，这里可以删掉，不把outputs重新变成元组。t5模型相同。

File: libai/models/t5_model.py [hunk 10]
```diff
@@ -0,0 +1,443 @@
...
+
+        self.position_ids = flow.arange(
+            max_seq_length,
+            dtype=flow.long,
+            sbp=dist.get_nd_sbp([flow.sbp.broadcast, flow.sbp.broadcast]),
+            placement=dist.get_layer_placement(layer_idx),
+        ).unsqueeze(0)
+
+    def forward(self, input_ids, past_length=0):
+        bsz, seq_length = input_ids.size()
```
Comment (dangkai4u) [inline libai/models/t5_model.py:63]: past_length是为了推理阶段使用，推理时past_key_values保存了之前的计算结果，此时的输入input_ids是一个token。通过past_key_values得到past_length，然后计算当前的input_ids的position_ids，需要根据past_length得到计算。

File: libai/models/t5_model.py [hunk 11]
```diff
@@ -0,0 +1,443 @@
...
+        return input_embeds
+
+    def word_embeddings(self):
+        return self.token_embeddings.weight
+
+
+class T5Encoder(nn.Module):
+    def __init__(
+        self,
+        embeddings,
```
Comment (dangkai4u) [inline libai/models/t5_model.py:81]: t5是自编码器，encoder和decoder共享了embedding模块，因此将embedding模块作为参数传到encoder和decoder模块中。

File: libai/models/t5_model.py [hunk 12]
```diff
@@ -0,0 +1,443 @@
...
+            past_length = 0
+        else:
+            past_length = past_key_values[0][0].size[2]
+
+        extended_attention_mask = self.extend_mask(attention_mask)
+        extended_attention_mask = self.casual_mask(
+            input_ids, past_length=past_length, attention_mask=extended_attention_mask
+        )
+
+        extended_encoder_attention_mask = self.extend_mask(encoder_attention_mask)
```
Comment (dangkai4u) [inline libai/models/t5_model.py:211]: 有的模型库选择在attention模块或transformer_layer里拓展attention_mask维度，这里选择统一拓展维度，然后传参到transformer_layer和attention中。区别在于：他们的做法经过了多次attention_mask拓展，我们经过了一次。我们的数据通信量和他们相比，可能会有略微增加（它们的attention_mask维度小，或动态构造，传输量少）。不确定哪种方法销量更高。

File: libai/models/t5_model.py [hunk 13]
```diff
@@ -0,0 +1,443 @@
...
+
+        self.embeddings = T5Embedding(
+            vocab_size,
+            hidden_size,
+            max_seq_length,
+            embedding_dropout_prob=embedding_dropout_prob,
+            init_method=init_method_normal(initializer_range),
+        )
+
+        self.encoder = T5Encoder(
```
Comment (dangkai4u) [inline libai/models/t5_model.py:273]: 这里分别构造了T5Encoder和T5Decoder，是因为推理阶段，需要先调用encoder，得到encoder_states，然后循环调用decoder，生成单词。推理阶段不会采用T5Model(xxx)方式调用模型。

File: libai/models/t5_model.py [hunk 14]
```diff
@@ -0,0 +1,443 @@
...
+
+        extended_encoder_attention_mask = self.extend_mask(encoder_attention_mask)
+
+        for i, (layer, past) in enumerate(zip(self.layers, past_key_values)):
+            if self.training:
+                hidden_states = layer(
+                    hidden_states,
+                    attention_mask=extended_attention_mask,
+                    encoder_states=encoder_states,
+                    encoder_attention_mask=extended_encoder_attention_mask,
```
Comment (dangkai4u) [inline libai/models/t5_model.py:219]: attention_mask是对输入（hidden_states）的mask矩阵，因此调用extended_attention_mask，在t5model的forward中对应decoder_attention_mask。encoder_attention_mask是对encoder_states的mask矩阵，这个mask矩阵和encoder的attention_mask相同，在t5model的forward中对应attention_mask。

File: libai/models/t5_model.py [hunk 15]
```diff
@@ -0,0 +1,443 @@
...
+    @staticmethod
+    def set_pipeline_stage_id(model):
+        dist_utils = dist.get_dist_util()
+
+        # Set pipeline parallelism stage_id
+        for module_block in model.modules():
+            # module.origin can get the original module
+            if isinstance(module_block.origin, T5Embedding):
+                module_block.config.stage_id = dist_utils.get_layer_stage_id(0)
+            elif isinstance(module_block.origin, (ExtendedMask, CasualMask)):
```
Comment (dangkai4u) [inline libai/models/t5_model.py:431]: 由于embedding共享，位于第一层，ExtendedMask和CasualMask也在模型最一开始调用，然后分配到不同的placement，因此位于layer0

File: libai/models/t5_model.py [hunk 16]
```diff
@@ -0,0 +1,443 @@
...
+                module_block.config.stage_id = dist_utils.get_layer_stage_id(0)
+            elif isinstance(module_block.origin, TransformerLayer):
+                module_block.config.stage_id = dist_utils.get_layer_stage_id(module_block.layer_idx)
+            elif isinstance(module_block.origin, LMLogits):
+                module_block.config.stage_id = dist_utils.get_layer_stage_id(-1)
+            elif isinstance(module_block.origin, ParallelCrossEntropyLoss):
+                module_block.config.stage_id = dist_utils.get_layer_stage_id(-1)
+
+        model.t5_model.encoder.layernorm_f.config.stage_id = dist_utils.get_layer_stage_id(
```
Comment (dangkai4u) [inline libai/models/t5_model.py:440]: encoder和decoder均包含一个layernorm层，encoder的layernorm应当放在encoder.num_layers-1，也就是encoder的最后一层，decoder的layernorm放在整个模型的最后一层。

File: libai/models/t5_model.py [hunk 17]
```diff
@@ -0,0 +1,443 @@
...
+                apply_query_key_layer_scaling=apply_query_key_layer_scaling,
+                layer_idx=layer_number,
+            )
+
+        self.embeddings = embeddings
+        self.extend_mask = ExtendedMask()
+        self.casual_mask = CasualMask()
+
+        self.layers = nn.ModuleList(
+            [build_layer(i + num_encoder_layers) for i in range(self.num_layers)]
```
Comment (dangkai4u) [inline libai/models/t5_model.py:184]: 这里要把num_encoder_layers作为参数输入过来，num_layers是解码器的层数，通过i+num_encoder_layers才能得到解码器每一层在整个模型中的layer_idx
